rfcs/0154-Migrate-taskcluster-to-postgres.md

```diff
@@ -0,0 +1,165 @@
+
+#### Data Volume
+
+To test the scalability and performance of the system, we will do an import of
+the FirefoxCI production database (minus the secrets) into the postgres database
+on the staging deployment and then observe if the database crashes or if there
+are any noticeable performance issues that arise. To make sure the database
+performs as expected when it has production quantities in it, we will initiate
+parallel requests on the same instance that we've importated the production
+quantities of data into. More on that in the next section.
+
+#### Parallel Requests
+
+To simulate, as closely as possible, actual traffic, we will create a script
+that initiates a big amount of processes (e.g., 10,000) where each process
+will simulate a number of parallel requests (e.g., 100) on the imported
+production quantities doing things like:
+* claimWork (an endpoint that interacts with a bunch of tables
+an queues)
+* create a worker type
+* resolve artifact
+* call the worker manager register endpoint
+* call often used endpoints like getTasks, getTaskGroup, getTaskByIndexRoute, etc.
+
+## Schema Migration
+
+Taskcluster will handle schema migrations and will have
+all migrations tested with sample data to ensure we get the expected schema. In
+addition, there will be one schema for the whole monorepo, so that schema
+updates happen atomically across all services.
+
+## Backups and Restores
+
+Teams operating Taskcluster will rely on the cloud provider's backup system to
+handle backups and restores.
+
+## DB Version Upgrades Compatibility
+
+When deploying a new version of Taskcluster, any necessary database upgrades
+*must* be applied before any services are upgraded.  Taskcluster will maintain
+the invariant that services expecting database version S can interoperate with a
+database at version V as long as V >= S.
+
+This invariant is maintained through careful attention to the definitions of the
+stored procedures (user-defined functions). These stored procedures are changed
+as necessary during the upgrade process, such that a procedure operates
+identically before and after the ugprade: identical arguments, identical return
+type, and correct behavior.
+
+_Example 1:_ If an upgrade factors a single table into two tables, then a
+procedure which previously queried from the single table would be rewritten
+during the ugprade to perform a join over the two new tables, returning
+appropriate results in the same format as before the upgrade.
+
+_Example 2:_ If you have a query that returns column foo, but in the next
+version your code wants to know the value of column bar as well, then a new
+stored procedure will need to be introduced to return both columns and stop
+using the existing stored procedure that returned the single column. That new
+stored procedure is then deployed before the code that uses it is deployed.
+
+A consequence of this design is that "procedures are forever" -- an upgrade can
```

Comment:
I think we can delete a procedure but it has to happen in a later upgrade, right?

Reply:
We also talked about supporting rollbacks during the all-hands. Can we add a section here talking about them and how we will support them or justifying why we won't.

Reply:
> I think we can delete a procedure but it has to happen in a later upgrade, right?

Rather than delete a procedure, a safer alternative would be to change the body of the function to return an empty array. `

> We also talked about supporting rollbacks during the all-hands. Can we add a section here talking about them and how we will support them or justifying why we won't.

As long as the procedure signature is not changed, rolling back shouldn't cause any issues.

> Can we add a section here talking about them

Will do.